dbms/src/Flash/Coprocessor/DAGContext.h

```diff
@@ -303,10 +302,11 @@ class DAGContext
 
     bool columnsForTestEmpty() { return columns_for_test_map.empty(); }
 
-    void cancelAllExchangeReceiver();
-
-    void initExchangeReceiverIfMPP(Context & context, size_t max_streams);
     const std::unordered_map<String, std::shared_ptr<ExchangeReceiver>> & getMPPExchangeReceiverMap() const;
+    void setMPPExchangeReceiverMap(std::unordered_map<String, std::shared_ptr<ExchangeReceiver>> * exchange_receiver_map)
+    {
+        mpp_exchange_receiver_map = exchange_receiver_map;
+    }
```

Comment:
`MPPTask` and `DAGContext` who destructs first?
Is there a risk of illegal pointers here?

Reply:
`mpp_exchange_receiver_map` is only used during compile, so it works fine for current code, but indeed it is error prone, I will use shared_ptr instead.